lib/elasticsearch.js

```diff
@@ -0,0 +1,593 @@
+'use strict';
+
+const AWS = require('aws-sdk');
+const AWSESConnection = require('http-aws-es');
+
+const { URL } = require('url');
+
+const UUID = require('uuid/v4');
+
+const elasticsearch = require('@elastic/elasticsearch');
+
+const ElasticSearchFilters = require('./elasticsearch-filters');
+const ElasticSearchError = require('./elasticsearch-error');
+
+const DEFAULT_LIMIT = 500;
+const ES_DEFAULT_HOST = 'http://localhost';
+
+/**
+ * @class ElasticSearch
+ * @classdesc Elasticsearch driver module
+*/
+class ElasticSearch {
+
+	constructor(config = {}) {
+		if(typeof config !== 'object' || Array.isArray(config))
+			throw new ElasticSearchError('Invalid config', ElasticSearchError.codes.INVALID_CONFIG);
+
+		this.config = {
+			protocol: config.protocol || '',
+			host: config.host || ES_DEFAULT_HOST,
+			port: config.port || '',
+			user: config.user || '',
+			password: config.password || '',
+			limit: config.limit || DEFAULT_LIMIT,
+			awsCredentials: config.awsCredentials || false
+		};
+
+		try {
+			const hostProtocol = `${new URL(this.config.host).protocol}//`; // throws if the URL doesn't have a protocol
+			this.config.protocol = hostProtocol;
+		} catch(err) {
+			// do nothing...
+		}
+
+		this.config.host = this.config.host.replace(this.config.protocol, ''); // Avoid protocol duplication
+	}
+
+	get _awsCredentials() {
+
+		if(!this.config.awsCredentials)
+			return {};
+
+		return {
+			connectionClass: AWSESConnection,
+			amazonES: { credentials: new AWS.EnvironmentCredentials('AWS') }
+		};
+	}
+
+	/**
+	 * Elasticsearch connection URL prefix
+	 * @returns {String} Elasticsearch URL prefix
+	*/
+	get _userPrefix() {
+		return this.config.user ? `${this.config.user}:${this.config.password}@` : '';
+	}
+
+	/**
+	 * Checks the connection to the Elasticsearch Client
+	 * @returns {elasticsearch.Client} Elasticsearch client with the connection
+	 * @throws if the connection is not successfull
+	*/
+	get client() {
+
+		if(!this._client) {
+			try {
+				this._client = new elasticsearch.Client({
+					node: `${this.config.protocol}${this._userPrefix}${this.config.host}${this.config.port ? `:${this.config.port}` : ''}`,
+					...this._awsCredentials
+				});
+			} catch(err) {
+				throw new ElasticSearchError(err.message, ElasticSearchError.codes.ELASTICSEARCH_ERROR);
+			}
+		}
+
+		return this._client;
+	}
+
+	/**
+	 * Validates the model structure
+	 * @param {Model} model Model instance
+	 * @throws If the model is invalid
+	 */
+	_validateModel(model) {
+
+		if(!model)
+			throw new ElasticSearchError('Empty model', ElasticSearchError.codes.INVALID_MODEL);
+
+		if(!model.constructor.table)
+			throw new ElasticSearchError('Invalid model', ElasticSearchError.codes.INVALID_MODEL);
+
+		if(typeof model.constructor.sortableFields !== 'object' || Array.isArray(model.constructor.sortableFields)) {
```

Comment:
Si no quisiera poner `fields` para ordenar,  estaria obligado a incluir `sortableFields` retornando vacio.
Seria buena idea que solo chequeo esto si existe.
```suggestion
		if(model.constructor.sortableFields && typeof model.constructor.sortableFields !== 'object' || Array.isArray(model.constructor.sortableFields)) {
```